Compiliation fails on ARM platforms (invalid type conversion)

Compilation for ARM platforms using gcc 7.2 fails with `invalid conversion from 'signed char*' to 'const char*'`. I believe this is because on ARM the char data type is unsigned by default, but in `src/formats/pngformat.cpp` and `src/formats/yasaraformat.cpp` negative values are assigned. I used Open Babel version 2.4.1 but it’s still the same in the latest source files.

The following patch which replaces the negative integer with its two’s complement works, but I have no idea if that breaks something else. The best solution would probably be to use `singed char`.

```
--- src/formats/pngformat.cpp.orig      2016-09-21 21:55:37.000000000 +0200
+++ src/formats/pngformat.cpp   2017-11-10 08:17:28.625719045 +0100
@@ -218,7 +218,7 @@
     _count=0;
     _hasInputPngFile=true;
   }
-  const char pngheader[] = {-119,80,78,71,13,10,26,10,0};
+  const char pngheader[] = {0x9,80,78,71,13,10,26,10,0};
   char readbytes[9];
   ifs.read(readbytes, 8);

--- src/formats/yasaraformat.cpp.orig   2016-09-21 21:55:37.000000000 +0200
+++ src/formats/yasaraformat.cpp        2017-11-10 08:07:02.620775489 +0100
@@ -472,7 +472,7 @@

   //  bool hetatom;
   char buffer[32],/*resname[4],*/atomname[5];
-  char double1[8]={0,0,0,0,0,0,-16,0x3f};
+  char double1[8]={0,0,0,0,0,0,0x10,0x3f};
   //   char *str;
   int i,j,/*m,q,*/pos;
   int /*resno,chainNum,link,linktype,*/atoms,element,links/*,chain*/;
```

Note to self: to force at least a warning message on an Intel machine, you can use "-funsigned-char -W -Wall":
```
/home/noel/Tools/openbabel-baoilleach2/src/formats/pngformat.cpp:221:58: warning: narrowing conversion of '-119' from 'int' to 'const char' inside { } is ill-formed in C++11 [-Wnarrowing]
   const char pngheader[]   = {-119,80,78,71,13,10,26,10,0};
```
The fix will be to use positive values and "unsigned char" as the type. A patch is on the way. Thanks for reporting.

--- a/src/formats/pngformat.cpp
+++ b/src/formats/pngformat.cpp
@@ -218,7 +218,7 @@ bool PNGFormat::ReadMolecule(OBBase* pOb, OBConversion* pConv)
     _count=0;
     _hasInputPngFile=true;
   }
-  const char pngheader[] = {-119,80,78,71,13,10,26,10,0};
+  const unsigned char pngheader[] = {137,80,78,71,13,10,26,10,0};
   char readbytes[9];
   ifs.read(readbytes, 8);
 

--- a/src/formats/yasaraformat.cpp
+++ b/src/formats/yasaraformat.cpp
@@ -472,7 +472,7 @@ bool YOBFormat::WriteMolecule(OBBase* pOb, OBConversion* pConv)
 
   //  bool hetatom;
   char buffer[32],/*resname[4],*/atomname[5];
-  char double1[8]={0,0,0,0,0,0,-16,0x3f};
+  unsigned char double1[8]={0,0,0,0,0,0,0xf0,0x3f};
   //   char *str;
   int i,j,/*m,q,*/pos;
   int /*resno,chainNum,link,linktype,*/atoms,element,links/*,chain*/;

--- a/src/formats/yasaraformat.cpp
+++ b/src/formats/yasaraformat.cpp
@@ -500,7 +500,7 @@ bool YOBFormat::WriteMolecule(OBBase* pOb, OBConversion* pConv)
   mem_set(buffer,0,8);
   for (i=0;i<4;i++)
   { for (j=0;j<4;j++)
-    { if (i==j) ofs.write(double1,8);
+    { if (i==j) ofs.write((char*)double1,8);
       else ofs.write(buffer,8); } }
   storeint32le(buffer,MOB_INFOEND);
   storeint32le(&buffer[4],MOB_INFOENDSIZE);

